### PR TITLE
Mistral-NeMo-12B recipe

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/__init__.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
     LlamaConfig,
     LlamaModel,
 )
-from nemo.collections.llm.gpt.model.mistral import MistralConfig7B, MistralModel
+from nemo.collections.llm.gpt.model.mistral import MistralConfig7B, MistralModel, MistralNeMo2407Config12B
```
Comment (ShriyaPalsamudram) [inline nemo/collections/llm/gpt/model/__init__.py:56]: is 2407 needed in the name? Can't we just use `MistralNeMoConfig12B`?

File: nemo/collections/llm/recipes/mistral_nemo_12b.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+from nemo.collections.llm.gpt.data.mock import MockDataModule
+from nemo.collections.llm.gpt.data.squad import SquadDataModule
+from nemo.collections.llm.gpt.model.mistral import MistralModel, MistralNeMo2407Config12B
+from nemo.collections.llm.peft.lora import LoRA
+from nemo.collections.llm.recipes.log.default import default_log, default_resume, tensorboard_logger
+from nemo.collections.llm.recipes.optim.adam import distributed_fused_adam_with_cosine_annealing
+from nemo.collections.llm.recipes.precision.mixed_precision import bf16_mixed
+from nemo.utils.exp_manager import TimingCallback
+
+NAME = "mistral_nemo_base_2407"
```
Comment (ShriyaPalsamudram) [inline nemo/collections/llm/recipes/mistral_nemo_12b.py:35]: Same as above, rename to `mistral_nemo_base`

File: nemo/collections/llm/recipes/mistral_nemo_12b.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        This is particularly useful for fine-tuning scenarios where you want to
+        start from the pre-trained Mistral-Nemo-Base-2407 model.
+    """
+    return run.Config(
+        nl.AutoResume, restore_config=run.Config(nl.RestoreConfig, path="hf://mistralai/Mistral-Nemo-Base-2407")
+    )
+
+
+@run.cli.factory(target=finetune, name=NAME)
+def finetune_recipe(
```
Comment (ShriyaPalsamudram) [inline nemo/collections/llm/recipes/mistral_nemo_12b.py:212]: Is this tested? If not, we should drop the finetune_recipe section and just keep pretrain for now.

File: nemo/collections/llm/gpt/model/__init__.py (hunk 1 above)
Comment (akoumpa) [reply nemo/collections/llm/gpt/model/__init__.py:56]: There's two names I think; https://huggingface.co/nvidia/Mistral-NeMo-12B-Base/tree/main but this only contains the nemo checkpoint and https://huggingface.co/mistralai/Mistral-Nemo-Base-2407 which contains the HF checkpoint.

File: nemo/collections/llm/recipes/mistral_nemo_12b.py (hunk 3 above)
Comment (akoumpa) [reply nemo/collections/llm/recipes/mistral_nemo_12b.py]: Thank you Shriya.

**Review by ShriyaPalsamudram — APPROVED** [review]

**Review by dimapihtar — DISMISSED** [review]
LGTM. Thank you!

**Review by dimapihtar — DISMISSED** [review]
LGTM. Thank you!

**Review by dimapihtar — APPROVED** [review]

**Review by dimapihtar — APPROVED** [review]
